### PR TITLE
Move object diff and field path from apimachinery

### PR DESCRIPTION
**Pull request by sttts**
Both packages are not apimachinery specific, but totally generic (and e.g. useful in k/kube-openapi).

### CODE REVIEW

File: diff/diff.go [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+func ObjectGoPrintDiff(a, b interface{}) string {
+	s := spew.ConfigState{DisableMethods: true}
+	return StringDiff(
+		s.Sprintf("%#v", a),
+		s.Sprintf("%#v", b),
+	)
+}
+
+// ObjectReflectDiff returns a diff computed through reflection, without serializing to JSON.
+func ObjectReflectDiff(a, b interface{}) string {
```
Comment (liggitt) [inline diff/diff.go]: @smarterclayton will be happy to have this more available

**Comment by liggitt** [comment]
both make sense to me for use outside k/k

File: diff/diff.go (hunk 1 above)
Comment (rajathagasthya) [reply diff/diff.go]: So am I! This fixes my request in #75.

File: diff/diff_test.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+		out  string
+	}{
+		"map": {
+			a: map[string]int{},
+			b: map[string]int{},
+		},
+		"detect nil map": {
+			a: map[string]int(nil),
+			b: map[string]int{},
+			out: `
```
Comment (thockin) [inline diff/diff_test.go]: this is a great place to use dedent and not have wacky indenting in string literals.

Comment (dims) [reply diff/diff_test.go]: @sttts do you want to switch to dedent in this PR or a follow up?

Comment (sttts) [reply diff/diff_test.go]: Would prefer a follow-up. This is a straight move.

**Comment by sttts** [comment]
> The fact that the diff APIs expose their implementation details in their names means we can probably never change these once we publish them. Are we cool with that?

I have my doubts those details should be hidden. It makes a big difference how you extract a diff from objects. JSON is different from reflect.

**Comment by thockin** [comment]
Perhaps the names should be more meaningfu, rather than
implementation-centric?  Or perhaps we should find one-true format that we
can use everywhere.

**Pull request merged**